dashboard/app/controllers/featured_projects_controller.rb

```diff
@@ -1,23 +1,35 @@
 class FeaturedProjectsController < ApplicationController
+  # The first time a project is featured.
   def create
-    unless current_user.try(:permission?, UserPermission::PROJECT_VALIDATOR)
+    if current_user.try(:permission?, UserPermission::PROJECT_VALIDATOR)
       _, channel_id = storage_decrypt_channel_id(featured_project_params[:project_id])
-      @featured_project = FeaturedProject.create({storage_app_id: channel_id, who_featured_user_id: current_user.id})
+      @featured_project = FeaturedProject.create({storage_app_id: channel_id, featured_at: DateTime.now, unfeatured_at: nil})
+      puts "featured project in create #{@featured_project}"
       @featured_project.save!
     end
   end
 
-  def destroy_by_project_id
-    unless current_user.try(:permission?, UserPermission::PROJECT_VALIDATOR)
-      _, channel_id = storage_decrypt_channel_id(params[:project_id])
-      @featured_project = FeaturedProject.find_by_storage_app_id(channel_id)
-      @featured_project.destroy
+  def unfeature
+    if current_user.try(:permission?, UserPermission::PROJECT_VALIDATOR)
+      _, channel_id = storage_decrypt_channel_id(featured_project_params[:project_id])
+      puts "channel_id > #{channel_id}"
+      @featured_project = FeaturedProject.find_by storage_app_id: channel_id
+      @featured_project.update_attribute(:unfeatured_at, DateTime.now)
+    end
+  end
+
+  def refeature
```

Comment:
it seems like it would be simpler, and less difficult to accidentally get into a bad state, if the "refeature" route was combined with "feature". They seem the same except that one uses "find_by" and the other uses "create". I think these can be combined roughly by starting with the logic you have here in "refeature", replacing "find_by" with "find_or_create_by".

the bad state we could accidentally get into would be if we did a "create" twice for the same project, perhaps if the project was open in two places.

as an added safeguard against ending up with two rows for the same project, `storage_app_id` should probably be unique in the database. apologies for missing that in the migration PR 😳 this seems ok to make in a separate PR after this one.

Reply:
+1 Good point. There isn't a need for both APIs. Feature should be smart enough to figure out whether it's new or re-featured.

Note [find_or_create_by!](https://apidock.com/rails/v4.0.2/ActiveRecord/Relation/find_or_create_by%21) can be supplied a block, which it will execute only on create (not when an existing entry is found). That may or may not be helpful here.